TiFlash crash after restart for Trying to add ref to non-exist page

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

<!-- a step by step guide for reproducing the bug. -->
consistency test and chaos tiflash failure
```
[2023/08/08 05:46:24.768 +08:00] [ERROR] [Exception.cpp:90] ["Code: 0, e.displayText() = DB::Exception: Trying to add ref to non-exist page [page_id=173.996173] [ori_id=173.14901] [ver=2312646.0] [resolve_id=0.0] [resolve_ver=0.0]:  [type=REF] [page_id=173.996173] [ver=2312646.0], e.what() = DB::Exception, Stack trace:\n\n\n       0x1c4b4ce\tDB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) [tiflash+29668558]\n                \tdbms/src/Common/Exception.h:46\n       0x1b8062c\tDB::PS::V3::PageDirectory<DB::PS::V3::u128::PageDirectoryTrait>::applyRefEditRecord(std::__1::map<DB::UInt128, std::__1::shared_ptr<DB::PS::V3::VersionedPageEntries<DB::PS::V3::u128::PageDirectoryTrait> >, std::__1::less<DB::UInt128>, std::__1::allocator<std::__1::pair<DB::UInt128 const, std::__1::shared_ptr<DB::PS::V3::VersionedPageEntries<DB::PS::V3::u128::PageDirectoryTrait> > > > >&, std::__1::shared_ptr<DB::PS::V3::VersionedPageEntries<DB::PS::V3::u128::PageDirectoryTrait> > const&, DB::PS::V3::PageEntriesEdit<DB::UInt128>::EditRecord const&, DB::PS::V3::PageVersion const&) [tiflash+28837420]\n                \tdbms/src/Storages/Page/V3/PageDirectory.cpp:1421\n       0x1b95bac\tDB::PS::V3::PageDirectoryFactory<DB::PS::V3::u128::FactoryTrait>::create(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::shared_ptr<DB::FileProvider>&, std::__1::shared_ptr<DB::PSDiskDelegator>&, DB::PS::V3::WALConfig const&) [tiflash+28924844]\n                \tdbms/src/Storages/Page/V3/PageDirectoryFactory.cpp:40\n       0x761ab42\tDB::PS::V3::PageStorageImpl::restore() [tiflash+123841346]\n                \tdbms/src/Storages/Page/V3/PageStorageImpl.cpp:80\n       0x749192b\tDB::DM::GlobalStoragePool::restore() [tiflash+122231083]\n                \tdbms/src/Storages/DeltaMerge/StoragePool.cpp:116\n       0x77d9091\tDB::Context::initializeGlobalStoragePoolIfNeed(DB::PathPool const&) [tiflash+125669521]\n                \tdbms/src/Interpreters/Context.cpp:1688\n       0x1cfd749\tDB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&) [tiflash+30398281]\n                \tdbms/src/Server/Server.cpp:1262\n       0x8a27d6a\tPoco::Util::Application::run() [tiflash+144866666]\n                \tcontrib/poco/Util/src/Application.cpp:335\n       0x1cf2dca\tDB::Server::run() [tiflash+30354890]\n                \tdbms/src/Server/Server.cpp:265\n       0x8a32654\tPoco::Util::ServerApplication::run(int, char**) [tiflash+144909908]\n                \tcontrib/poco/Util/src/ServerApplication.cpp:618\n       0x1d0c9b9\tmainEntryClickHouseServer(int, char**) [tiflash+30460345]\n                \tdbms/src/Server/Server.cpp:1742\n       0x1c378ec\tmain [tiflash+29587692]\n                \tdbms/src/Server/main.cpp:167\n  0x7f5ebf3e2eb0\t__libc_start_call_main [libc.so.6+261808]\n  0x7f5ebf3e2f60\t__libc_start_main_impl [libc.so.6+261984]\n       0x19274a9\t<unknown symbol> [tiflash+26375337]"] [source="bool DB::Context::initializeGlobalStoragePoolIfNeed(const DB::PathPool &)"] [thread_id=1]
[2023/08/08 05:46:25.040 +08:00] [ERROR] [<unknown>] ["DB::Exception: Trying to add ref to non-exist page [page_id=173.996173] [ori_id=173.14901] [ver=2312646.0] [resolve_id=0.0] [resolve_ver=0.0]:  [type=REF] [page_id=173.996173] [ver=2312646.0]"] [source=Application] [thread_id=1]
[2023/08/08 05:51:36.825 +08:00] [ERROR] [Exception.cpp:90] ["Code: 0, e.displayText() = DB::Exception: Trying to add ref to non-exist page [page_id=173.996173] [ori_id=173.14901] [ver=2312646.0] [resolve_id=0.0] [resolve_ver=0.0]:  [type=REF] [page_id=173.996173] [ver=2312646.0], e.what() = DB::Exception, Stack trace:\n\n\n       0x1c4b4ce\tDB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) [tiflash+29668558]\n                \tdbms/src/Common/Exception.h:46\n       0x1b8062c\tDB::PS::V3::PageDirectory<DB::PS::V3::u128::PageDirectoryTrait>::applyRefEditRecord(std::__1::map<DB::UInt128, std::__1::shared_ptr<DB::PS::V3::VersionedPageEntries<DB::PS::V3::u128::PageDirectoryTrait> >, std::__1::less<DB::UInt128>, std::__1::allocator<std::__1::pair<DB::UInt128 const, std::__1::shared_ptr<DB::PS::V3::VersionedPageEntries<DB::PS::V3::u128::PageDirectoryTrait> > > > >&, std::__1::shared_ptr<DB::PS::V3::VersionedPageEntries<DB::PS::V3::u128::PageDirectoryTrait> > const&, DB::PS::V3::PageEntriesEdit<DB::UInt128>::EditRecord const&, DB::PS::V3::PageVersion const&) [tiflash+28837420]\n                \tdbms/src/Storages/Page/V3/PageDirectory.cpp:1421\n       0x1b95bac\tDB::PS::V3::PageDirectoryFactory<DB::PS::V3::u128::FactoryTrait>::create(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::shared_ptr<DB::FileProvider>&, std::__1::shared_ptr<DB::PSDiskDelegator>&, DB::PS::V3::WALConfig const&) [tiflash+28924844]\n                \tdbms/src/Storages/Page/V3/PageDirectoryFactory.cpp:40\n       0x761ab42\tDB::PS::V3::PageStorageImpl::restore() [tiflash+123841346]\n                \tdbms/src/Storages/Page/V3/PageStorageImpl.cpp:80\n       0x749192b\tDB::DM::GlobalStoragePool::restore() [tiflash+122231083]\n                \tdbms/src/Storages/DeltaMerge/StoragePool.cpp:116\n       0x77d9091\tDB::Context::initializeGlobalStoragePoolIfNeed(DB::PathPool const&) [tiflash+125669521]\n                \tdbms/src/Interpreters/Context.cpp:1688\n       0x1cfd749\tDB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&) [tiflash+30398281]\n                \tdbms/src/Server/Server.cpp:1262\n       0x8a27d6a\tPoco::Util::Application::run() [tiflash+144866666]\n                \tcontrib/poco/Util/src/Application.cpp:335\n       0x1cf2dca\tDB::Server::run() [tiflash+30354890]\n                \tdbms/src/Server/Server.cpp:265\n       0x8a32654\tPoco::Util::ServerApplication::run(int, char**) [tiflash+144909908]\n                \tcontrib/poco/Util/src/ServerApplication.cpp:618\n       0x1d0c9b9\tmainEntryClickHouseServer(int, char**) [tiflash+30460345]\n                \tdbms/src/Server/Server.cpp:1742\n       0x1c378ec\tmain [tiflash+29587692]\n                \tdbms/src/Server/main.cpp:167\n  0x7f4d99c07eb0\t__libc_start_call_main [libc.so.6+261808]\n  0x7f4d99c07f60\t__libc_start_main_impl [libc.so.6+261984]\n       0x19274a9\t<unknown symbol> [tiflash+26375337]"] [source="bool DB::Context::initializeGlobalStoragePoolIfNeed(const DB::PathPool &)"] [thread_id=1]
[2023/08/08 05:51:37.113 +08:00] [ERROR] [<unknown>] ["DB::Exception: Trying to add ref to non-exist page [page_id=173.996173] [ori_id=173.14901] [ver=2312646.0] [resolve_id=0.0] [resolve_ver=0.0]:  [type=REF] [page_id=173.996173] [ver=2312646.0]"] [source=Application] [thread_id=1]
[2023/08/08 05:56:47.823 +08:00] [ERROR] [Exception.cpp:90] ["Code: 0, e.displayText() = DB::Exception: Trying to add ref to non-exist page [page_id=173.996173] [ori_id=173.14901] [ver=2312646.0] [resolve_id=0.0] [resolve_ver=0.0]:  [type=REF] [page_id=173.996173] [ver=2312646.0], e.what() = DB::Exception, Stack trace:\n\n\n       0x1c4b4ce\tDB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) [tiflash+29668558]\n                \tdbms/src/Common/Exception.h:46\n       0x1b8062c\tDB::PS::V3::PageDirectory<DB::PS::V3::u128::PageDirectoryTrait>::applyRefEditRecord(std::__1::map<DB::UInt128, std::__1::shared_ptr<DB::PS::V3::VersionedPageEntries<DB::PS::V3::u128::PageDirectoryTrait> >, std::__1::less<DB::UInt128>, std::__1::allocator<std::__1::pair<DB::UInt128 const, std::__1::shared_ptr<DB::PS::V3::VersionedPageEntries<DB::PS::V3::u128::PageDirectoryTrait> > > > >&, std::__1::shared_ptr<DB::PS::V3::VersionedPageEntries<DB::PS::V3::u128::PageDirectoryTrait> > const&, DB::PS::V3::PageEntriesEdit<DB::UInt128>::EditRecord const&, DB::PS::V3::PageVersion const&) [tiflash+28837420]\n                \tdbms/src/Storages/Page/V3/PageDirectory.cpp:1421\n       0x1b95bac\tDB::PS::V3::PageDirectoryFactory<DB::PS::V3::u128::FactoryTrait>::create(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::shared_ptr<DB::FileProvider>&, std::__1::shared_ptr<DB::PSDiskDelegator>&, DB::PS::V3::WALConfig const&) [tiflash+28924844]\n                \tdbms/src/Storages/Page/V3/PageDirectoryFactory.cpp:40\n       0x761ab42\tDB::PS::V3::PageStorageImpl::restore() [tiflash+123841346]\n                \tdbms/src/Storages/Page/V3/PageStorageImpl.cpp:80\n       0x749192b\tDB::DM::GlobalStoragePool::restore() [tiflash+122231083]\n                \tdbms/src/Storages/DeltaMerge/StoragePool.cpp:116\n       0x77d9091\tDB::Context::initializeGlobalStoragePoolIfNeed(DB::PathPool const&) [tiflash+125669521]\n                \tdbms/src/Interpreters/Context.cpp:1688\n       0x1cfd749\tDB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&) [tiflash+30398281]\n                \tdbms/src/Server/Server.cpp:1262\n       0x8a27d6a\tPoco::Util::Application::run() [tiflash+144866666]\n                \tcontrib/poco/Util/src/Application.cpp:335\n       0x1cf2dca\tDB::Server::run() [tiflash+30354890]\n                \tdbms/src/Server/Server.cpp:265\n       0x8a32654\tPoco::Util::ServerApplication::run(int, char**) [tiflash+144909908]\n                \tcontrib/poco/Util/src/ServerApplication.cpp:618\n       0x1d0c9b9\tmainEntryClickHouseServer(int, char**) [tiflash+30460345]\n                \tdbms/src/Server/Server.cpp:1742\n       0x1c378ec\tmain [tiflash+29587692]\n                \tdbms/src/Server/main.cpp:167\n  0x7f0bb4cabeb0\t__libc_start_call_main [libc.so.6+261808]\n  0x7f0bb4cabf60\t__libc_start_main_impl [libc.so.6+261984]\n       0x19274a9\t<unknown symbol> [tiflash+26375337]"] [source="bool DB::Context::initializeGlobalStoragePoolIfNeed(const DB::PathPool &)"] [thread_id=1]
[2023/08/08 05:56:47.924 +08:00] [ERROR] [<unknown>] ["DB::Exception: Trying to add ref to non-exist page [page_id=173.996173] [ori_id=173.14901] [ver=2312646.0] [resolve_id=0.0] [resolve_ver=0.0]:  [type=REF] [page_id=173.996173] [ver=2312646.0]"] [source=Application] [thread_id=1]
[2023/08/08 06:02:04.853 +08:00] [ERROR] [Exception.cpp:90] ["Code: 0, e.displayText() = DB::Exception: Trying to add ref to non-exist page [page_id=173.996173] [ori_id=173.14901] [ver=2312646.0] [resolve_id=0.0] [resolve_ver=0.0]:  [type=REF] [page_id=173.996173] [ver=2312646.0], e.what() = DB::Exception, Stack trace:\n\n\n       0x1c4b4ce\tDB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) [tiflash+29668558]\n                \tdbms/src/Common/Exception.h:46\n       0x1b8062c\tDB::PS::V3::PageDirectory<DB::PS::V3::u128::PageDirectoryTrait>::applyRefEditRecord(std::__1::map<DB::UInt128, std::__1::shared_ptr<DB::PS::V3::VersionedPageEntries<DB::PS::V3::u128::PageDirectoryTrait> >, std::__1::less<DB::UInt128>, std::__1::allocator<std::__1::pair<DB::UInt128 const, std::__1::shared_ptr<DB::PS::V3::VersionedPageEntries<DB::PS::V3::u128::PageDirectoryTrait> > > > >&, std::__1::shared_ptr<DB::PS::V3::VersionedPageEntries<DB::PS::V3::u128::PageDirectoryTrait> > const&, DB::PS::V3::PageEntriesEdit<DB::UInt128>::EditRecord const&, DB::PS::V3::PageVersion const&) [tiflash+28837420]\n                \tdbms/src/Storages/Page/V3/PageDirectory.cpp:1421\n       0x1b95bac\tDB::PS::V3::PageDirectoryFactory<DB::PS::V3::u128::FactoryTrait>::create(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::shared_ptr<DB::FileProvider>&, std::__1::shared_ptr<DB::PSDiskDelegator>&, DB::PS::V3::WALConfig const&) [tiflash+28924844]\n                \tdbms/src/Storages/Page/V3/PageDirectoryFactory.cpp:40\n       0x761ab42\tDB::PS::V3::PageStorageImpl::restore() [tiflash+123841346]\n                \tdbms/src/Storages/Page/V3/PageStorageImpl.cpp:80\n       0x749192b\tDB::DM::GlobalStoragePool::restore() [tiflash+122231083]\n                \tdbms/src/Storages/DeltaMerge/StoragePool.cpp:116\n       0x77d9091\tDB::Context::initializeGlobalStoragePoolIfNeed(DB::PathPool const&) [tiflash+125669521]\n                \tdbms/src/Interpreters/Context.cpp:1688\n       0x1cfd749\tDB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&) [tiflash+30398281]\n                \tdbms/src/Server/Server.cpp:1262\n       0x8a27d6a\tPoco::Util::Application::run() [tiflash+144866666]\n                \tcontrib/poco/Util/src/Application.cpp:335\n       0x1cf2dca\tDB::Server::run() [tiflash+30354890]\n                \tdbms/src/Server/Server.cpp:265\n       0x8a32654\tPoco::Util::ServerApplication::run(int, char**) [tiflash+144909908]\n                \tcontrib/poco/Util/src/ServerApplication.cpp:618\n       0x1d0c9b9\tmainEntryClickHouseServer(int, char**) [tiflash+30460345]\n                \tdbms/src/Server/Server.cpp:1742\n       0x1c378ec\tmain [tiflash+29587692]\n                \tdbms/src/Server/main.cpp:167\n  0x7fb8fd53eeb0\t__libc_start_call_main [libc.so.6+261808]\n  0x7fb8fd53ef60\t__libc_start_main_impl [libc.so.6+261984]\n       0x19274a9\t<unknown symbol> [tiflash+26375337]"] [source="bool DB::Context::initializeGlobalStoragePoolIfNeed(const DB::PathPool &)"] [thread_id=1]
[2023/08/08 06:02:05.048 +08:00] [ERROR] [<unknown>] ["DB::Exception: Trying to add ref to non-exist page [page_id=173.996173] [ori_id=173.14901] [ver=2312646.0] [resolve_id=0.0] [resolve_ver=0.0]:  [type=REF] [page_id=173.996173] [ver=2312646.0]"] [source=Application] [thread_id=1]
[2023/08/08 06:07:14.860 +08:00] [ERROR] [Exception.cpp:90] ["Code: 0, e.displayText() = DB::Exception: Trying to add ref to non-exist page [page_id=173.996173] [ori_id=173.14901] [ver=2312646.0] [resolve_id=0.0] [resolve_ver=0.0]:  [type=REF] [page_id=173.996173] [ver=2312646.0], e.what() = DB::Exception, Stack trace:\n\n\n       0x1c4b4ce\tDB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) [tiflash+29668558]\n                \tdbms/src/Common/Exception.h:46\n       0x1b8062c\tDB::PS::V3::PageDirectory<DB::PS::V3::u128::PageDirectoryTrait>::applyRefEditRecord(std::__1::map<DB::UInt128, std::__1::shared_ptr<DB::PS::V3::VersionedPageEntries<DB::PS::V3::u128::PageDirectoryTrait> >, std::__1::less<DB::UInt128>, std::__1::allocator<std::__1::pair<DB::UInt128 const, std::__1::shared_ptr<DB::PS::V3::VersionedPageEntries<DB::PS::V3::u128::PageDirectoryTrait> > > > >&, std::__1::shared_ptr<DB::PS::V3::VersionedPageEntries<DB::PS::V3::u128::PageDirectoryTrait> > const&, DB::PS::V3::PageEntriesEdit<DB::UInt128>::EditRecord const&, DB::PS::V3::PageVersion const&) [tiflash+28837420]\n                \tdbms/src/Storages/Page/V3/PageDirectory.cpp:1421\n       0x1b95bac\tDB::PS::V3::PageDirectoryFactory<DB::PS::V3::u128::FactoryTrait>::create(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::shared_ptr<DB::FileProvider>&, std::__1::shared_ptr<DB::PSDiskDelegator>&, DB::PS::V3::WALConfig const&) [tiflash+28924844]\n                \tdbms/src/Storages/Page/V3/PageDirectoryFactory.cpp:40\n       0x761ab42\tDB::PS::V3::PageStorageImpl::restore() [tiflash+123841346]\n                \tdbms/src/Storages/Page/V3/PageStorageImpl.cpp:80\n       0x749192b\tDB::DM::GlobalStoragePool::restore() [tiflash+122231083]\n                \tdbms/src/Storages/DeltaMerge/StoragePool.cpp:116\n       0x77d9091\tDB::Context::initializeGlobalStoragePoolIfNeed(DB::PathPool const&) [tiflash+125669521]\n                \tdbms/src/Interpreters/Context.cpp:1688\n       0x1cfd749\tDB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&) [tiflash+30398281]\n                \tdbms/src/Server/Server.cpp:1262\n       0x8a27d6a\tPoco::Util::Application::run() [tiflash+144866666]\n                \tcontrib/poco/Util/src/Application.cpp:335\n       0x1cf2dca\tDB::Server::run() [tiflash+30354890]\n                \tdbms/src/Server/Server.cpp:265\n       0x8a32654\tPoco::Util::ServerApplication::run(int, char**) [tiflash+144909908]\n                \tcontrib/poco/Util/src/ServerApplication.cpp:618\n       0x1d0c9b9\tmainEntryClickHouseServer(int, char**) [tiflash+30460345]\n                \tdbms/src/Server/Server.cpp:1742\n       0x1c378ec\tmain [tiflash+29587692]\n                \tdbms/src/Server/main.cpp:167\n  0x7fb04af61eb0\t__libc_start_call_main [libc.so.6+261808]\n  0x7fb04af61f60\t__libc_start_main_impl [libc.so.6+261984]\n       0x19274a9\t<unknown symbol> [tiflash+26375337]"] [source="bool DB::Context::initializeGlobalStoragePoolIfNeed(const DB::PathPool &)"] [thread_id=1]
```
### 2. What did you expect to see? (Required)

### 3. What did you see instead (Required)

### 4. What is your TiFlash version? (Required)
master

--- a/dbms/src/Storages/Page/V3/PageDirectoryFactory.cpp
+++ b/dbms/src/Storages/Page/V3/PageDirectoryFactory.cpp
@@ -346,6 +346,10 @@ void PageDirectoryFactory<Trait>::loadFromDisk(const PageDirectoryPtr & dir, WAL
 {
     DataFileIdSet data_file_ids;
     auto checkpoint_snap_seq = reader->getSnapSeqForCheckpoint();
+    // make sure the max sequence is larger than the checkpoint sequence
+    if (max_applied_ver.sequence < checkpoint_snap_seq)
+        max_applied_ver = PageVersion(checkpoint_snap_seq, 0);
+
     while (reader->remained())
     {
         auto [from_checkpoint, record] = reader->next();

--- a/dbms/src/Storages/Page/V3/PageDirectoryFactory.cpp
+++ b/dbms/src/Storages/Page/V3/PageDirectoryFactory.cpp
@@ -346,7 +346,7 @@ void PageDirectoryFactory<Trait>::loadFromDisk(const PageDirectoryPtr & dir, WAL
 {
     DataFileIdSet data_file_ids;
     auto checkpoint_snap_seq = reader->getSnapSeqForCheckpoint();
-    // make sure the max sequence is larger than the checkpoint sequence
+    // make sure the max sequence is larger or equal than the checkpoint sequence
     if (max_applied_ver.sequence < checkpoint_snap_seq)
         max_applied_ver = PageVersion(checkpoint_snap_seq, 0);
 

--- a/dbms/src/Storages/Page/V3/tests/gtest_page_directory.cpp
+++ b/dbms/src/Storages/Page/V3/tests/gtest_page_directory.cpp
@@ -1549,6 +1549,64 @@ try
 }
 CATCH
 
+TEST_F(PageDirectoryTest, Issue7915Case4)
+try
+{
+    PageEntryV3
+        entry_1_v1{.file_id = 50, .size = 7890, .padded_size = 0, .tag = 0, .offset = 0x123, .checksum = 0x4567};
+    PageEntryV3
+        entry_2_v1{.file_id = 50, .size = 7890, .padded_size = 0, .tag = 0, .offset = 0x123, .checksum = 0x4567};
+    {
+        PageEntriesEdit edit;
+        edit.put(buildV3Id(TEST_NAMESPACE_ID, 1), entry_1_v1);
+        edit.put(buildV3Id(TEST_NAMESPACE_ID, 2), entry_2_v1);
+        dir->apply(std::move(edit));
+    }
+
+    {
+        PageEntriesEdit edit;
+        edit.ref(buildV3Id(TEST_NAMESPACE_ID, 3), buildV3Id(TEST_NAMESPACE_ID, 2));
+        dir->apply(std::move(edit));
+    }
+    {
+        PageEntriesEdit edit;
+        edit.del(buildV3Id(TEST_NAMESPACE_ID, 2));
+        edit.del(buildV3Id(TEST_NAMESPACE_ID, 3));
+        dir->apply(std::move(edit));
+    }
+
+    {
+        dir->gcInMemEntries(u128::PageDirectoryType::InMemGCOption{.need_removed_entries = false});
+        ASSERT_TRUE(dir->tryDumpSnapshot(nullptr, true));
+    }
+
+    // restart and check
+    dir = restoreFromDisk();
+    {
+        EXPECT_EQ(dir->numPages(), 1);
+        auto snap = dir->createSnapshot("");
+        EXPECT_ENTRY_EQ(entry_1_v1, dir, 1, snap);
+    }
+
+    {
+        PageEntriesEdit edit;
+        edit.put(buildV3Id(TEST_NAMESPACE_ID, 10), entry_2_v1);
+        edit.ref(buildV3Id(TEST_NAMESPACE_ID, 11), buildV3Id(TEST_NAMESPACE_ID, 10));
+        dir->apply(std::move(edit));
+    }
+
+    // restart again
+    dir = restoreFromDisk();
+    {
+        EXPECT_EQ(dir->numPages(), 3);
+        auto snap = dir->createSnapshot("");
+        EXPECT_ENTRY_EQ(entry_1_v1, dir, 1, snap);
+        auto normal_id = getNormalPageIdU64(dir, 11, snap);
+        EXPECT_EQ(normal_id, 10);
+    }
+}
+CATCH
+
 TEST(MultiVersionRefCount, RefAndCollapse)
 try
 {

